File: server/src/main/java/mahjong/rules/Agari.java
package mahjong.rules;

import java.util.ArrayList;
import java.util.List;

import mahjong.core.Meld;
import mahjong.core.Tiles;

/** 和了形分解：把一个和了手牌枚举成所有可能的「4 面子 + 1 雀头 / 七对子 / 国士无双」解释。 */
public final class Agari {

    public static final int TYPE_STANDARD = 0;
    public static final int TYPE_CHIITOITSU = 1;
    public static final int TYPE_KOKUSHI = 2;

    public static final int SET_RUN = 0;
    public static final int SET_TRIPLET = 1;
    public static final int SET_QUAD = 2;

    public static final int WAIT_RYANMEN = 0;
    public static final int WAIT_KANCHAN = 1;
    public static final int WAIT_PENCHAN = 2;
    public static final int WAIT_SHANPON = 3;
    public static final int WAIT_TANKI = 4;

    private Agari() {
    }

    /** 一种和了形解释。 */
    public static final class Form {
        public int type = TYPE_STANDARD;
        public int pair = -1;
        public int nSets = 0;
        public final int[] setType = new int[4];
        public final int[] setStart = new int[4];
        public final boolean[] setConcealed = new boolean[4];
        public final boolean[] setFromMeld = new boolean[4];
        /** 和了牌所在的面子下标；-1 表示在雀头。 */
        public int winSet = -1;
        public int waitType = WAIT_RYANMEN;
        /** 国士无双十三面。 */
        public boolean kokushi13 = false;
        /** 供 Evaluator 填入副露引用。 */
        public final Meld[] setMeld = new Meld[4];



        public Form copy() {
            Form f = new Form();
            f.type = type;
            f.pair = pair;
            f.nSets = nSets;
            System.arraycopy(setType, 0, f.setType, 0, 4);
            System.arraycopy(setStart, 0, f.setStart, 0, 4);
            System.arraycopy(setConcealed, 0, f.setConcealed, 0, 4);
            System.arraycopy(setFromMeld, 0, f.setFromMeld, 0, 4);
            System.arraycopy(setMeld, 0, f.setMeld, 0, 4);
            f.winSet = winSet;
            f.waitType = waitType;
            f.kokushi13 = kokushi13;
            return f;
        }
    }

    /**
     * 分解和了形。
     *
     * @param counts    暗手牌计数（**含**和了牌）
     * @param melds     已副露的面子
     * @param winKind   和了牌的 kind
     * @return 所有合法解释（可能为空，表示不和了）
     */
    public static List<Form> decompose(int[] counts, List<Meld> melds, int winKind) {
        List<Form> out = new ArrayList<>();
        int meldCount = melds == null ? 0 : melds.size();
        int need = 4 - meldCount;
        int[] c = counts.clone();
        int total = Tiles.sum(c);
        if (total != need * 3 + 2) {
            return out;
        }
        // ---- 一般型
        List<int[]> rawShapes = new ArrayList<>(); // 编码：pair, nSets, [type,start]...
        for (int p = 0; p < Tiles.KIND_COUNT; p++) {
            if (c[p] < 2) {
                continue;
            }
            c[p] -= 2;
            Shape shape = new Shape();
            shape.pair = p;
            enumSets(c, 0, need, shape, rawShapes);
            c[p] += 2;
        }
        for (int[] shape : rawShapes) {
            addWinVariants(out, shape, winKind, melds);
        }
        // ---- 七对子
        if (meldCount == 0) {
            boolean ok = true;
            for (int k = 0; k < Tiles.KIND_COUNT; k++) {
                if (c[k] != 0 && c[k] != 2) {
                    ok = false;
                    break;
                }
            }
            if (ok) {
                Form f = new Form();
                f.type = TYPE_CHIITOITSU;
                f.pair = winKind;
                f.nSets = 0;
                f.winSet = -1;
                f.waitType = WAIT_TANKI;
                out.add(f);
            }
            // ---- 国士无双
            if (isKokushi(c)) {
                Form f = new Form();
                f.type = TYPE_KOKUSHI;
                f.pair = winKind;
                f.nSets = 0;
                int[] pre = c.clone();
                pre[winKind]--;
                f.kokushi13 = allOrphanKinds(pre);
                f.winSet = -1;
                f.waitType = WAIT_TANKI;
                out.add(f);
            }
        }
        return out;
    }

    private static boolean isKokushi(int[] c) {
        int kinds = 0;
        int total = 0;
        for (int k = 0; k < Tiles.KIND_COUNT; k++) {
            if (c[k] == 0) {
                continue;
            }
            if (!Tiles.isYaochu(k)) {
                return false;
            }
            kinds++;
            total += c[k];
            if (c[k] > 2) {
                return false;
            }
        }
        return kinds == 13 && total == 14;
    }

    private static boolean allOrphanKinds(int[] c) {
        int kinds = 0;
        for (int k = 0; k < Tiles.KIND_COUNT; k++) {
            if (Tiles.isYaochu(k) && c[k] > 0) {
                kinds++;
            }
        }
        return kinds == 13;
    }

    private static final class Shape {
        int pair;
        int nSets;
        final int[] type = new int[4];
        final int[] start = new int[4];
    }

    private static void enumSets(int[] c, int from, int need, Shape shape, List<int[]> out) {
        if (need == 0) {
            for (int k = from; k < Tiles.KIND_COUNT; k++) {
                if (c[k] != 0) {
                    return;
                }
            }
            int[] enc = new int[2 + shape.nSets * 2];
            enc[0] = shape.pair;
            enc[1] = shape.nSets;
            for (int i = 0; i < shape.nSets; i++) {
                enc[2 + i * 2] = shape.type[i];
                enc[3 + i * 2] = shape.start[i];
            }
            out.add(enc);
            return;
        }
        int i = from;
        while (i < Tiles.KIND_COUNT && c[i] == 0) {
            i++;
        }
        if (i == Tiles.KIND_COUNT) {
            return;
        }
        // 刻子
        if (c[i] >= 3) {
            c[i] -= 3;
            shape.type[shape.nSets] = SET_TRIPLET;
            shape.start[shape.nSets] = i;
            shape.nSets++;
            enumSets(c, i, need - 1, shape, out);
            shape.nSets--;
            c[i] += 3;
        }
        // 顺子
        if (i < 27 && i % 9 <= 6 && c[i + 1] > 0 && c[i + 2] > 0) {
            c[i]--;
            c[i + 1]--;
            c[i + 2]--;
            shape.type[shape.nSets] = SET_RUN;
            shape.start[shape.nSets] = i;
            shape.nSets++;
            enumSets(c, i, need - 1, shape, out);
            shape.nSets--;
            c[i]++;
            c[i + 1]++;
            c[i + 2]++;
        }
    }

    /** 为一个「形」加上副露面子，并展开和了牌归属的所有变体。 */
    private static void addWinVariants(List<Form> out, int[] enc, int winKind, List<Meld> melds) {
        int pair = enc[0];
        int n = enc[1];
        int meldCount = melds == null ? 0 : melds.size();
        int total4 = n + meldCount;
        if (total4 != 4) {
            return;
        }
        // 基础形（不含副露）
        Form base = new Form();
        base.pair = pair;
        base.nSets = n;
        for (int i = 0; i < n; i++) {
            base.setType[i] = enc[2 + i * 2];
            base.setStart[i] = enc[3 + i * 2];
            base.setConcealed[i] = true;
            base.setFromMeld[i] = false;
        }
        // 把和了牌归属展开
        List<int[]> assign = new ArrayList<>(); // [setIndex(-1=雀头), waitType]
        if (pair == winKind) {
            assign.add(new int[]{-1, WAIT_TANKI});
        }
        for (int i = 0; i < n; i++) {
            if (base.setType[i] == SET_RUN) {
                int s = base.setStart[i];
                if (winKind == s) {
                    // 缺顺子最小牌：手中剩 s+1,s+2；只有 8,9 才是边张
                    assign.add(new int[]{i, (s % 9 == 6) ? WAIT_PENCHAN : WAIT_RYANMEN});
                } else if (winKind == s + 1) {
                    assign.add(new int[]{i, WAIT_KANCHAN});
                } else if (winKind == s + 2) {
                    // 缺顺子最大牌：手中剩 s,s+1；只有 1,2 才是边张
                    assign.add(new int[]{i, (s % 9 == 0) ? WAIT_PENCHAN : WAIT_RYANMEN});
                }
            } else if (base.setType[i] == SET_TRIPLET && base.setStart[i] == winKind) {
                assign.add(new int[]{i, WAIT_SHANPON});
            }
        }
        if (assign.isEmpty()) {
            return;
        }
        for (int[] a : assign) {
            Form f = base.copy();
            f.winSet = a[0];
            f.waitType = a[1];
            // 追加副露面子
            int idx = f.nSets;
            for (Meld m : melds) {
                f.setType[idx] = m.isKan() ? SET_QUAD : (m.isRun() ? SET_RUN : SET_TRIPLET);
                f.setStart[idx] = m.baseKind();
                f.setConcealed[idx] = !m.open();
                f.setFromMeld[idx] = true;
                f.setMeld[idx] = m;
                idx++;
            }
            f.nSets = idx;
            out.add(f);
        }
    }

    /** 是否和了形（不检查役种）。 */
    public static boolean isAgari(int[] counts, int meldCount) {
        return Shanten.min(counts, meldCount) < 0;
    }

    /**
     * 计算听牌种类（返回 kind 列表）。
     *
     * @param counts13 暗手 13 张（或 13-3k 张）计数
     */
    public static List<Integer> waits(int[] counts13, int meldCount) {
        List<Integer> out = new ArrayList<>();
        int[] c = counts13.clone();
        for (int k = 0; k < Tiles.KIND_COUNT; k++) {
            if (c[k] >= 4) {
                continue;
            }
            c[k]++;
            if (isAgari(c, meldCount)) {
                out.add(k);
            }
            c[k]--;
        }
        return out;
    }
}

File: trainer/src/agari.cpp
#include "agari.hpp"

#include <vector>

#include "shanten.hpp"
#include "tiles.hpp"

namespace trainer {
namespace {

/** 枚举出的"暗手部分"形状：雀头 + n 个面子（不含副露）。 */
struct Shape {
    int pair = -1;
    int nSets = 0;
    std::array<int, 4> type{};
    std::array<int, 4> start{};
};

/** 与 Java `Agari.enumSets` 逐行同序（刻子在前、顺子在后；`need == 0` 要求恰好拆空）。 */
void enumSets(Counts &c, int from, int need, Shape &shape, std::vector<Shape> &out) {
    if (need == 0) {
        for (int k = from; k < kKindCount; k++) {
            if (c[static_cast<size_t>(k)] != 0) {
                return;
            }
        }
        out.push_back(shape);
        return;
    }
    int i = from;
    while (i < kKindCount && c[static_cast<size_t>(i)] == 0) {
        i++;
    }
    if (i == kKindCount) {
        return;
    }
    auto &ci = c[static_cast<size_t>(i)];
    if (ci >= 3) {                                       // 刻子
        ci -= 3;
        shape.type[static_cast<size_t>(shape.nSets)] = kSetTriplet;
        shape.start[static_cast<size_t>(shape.nSets)] = i;
        shape.nSets++;
        enumSets(c, i, need - 1, shape, out);
        shape.nSets--;
        ci += 3;
    }
    if (i < 27 && i % 9 <= 6 && c[static_cast<size_t>(i + 1)] > 0
            && c[static_cast<size_t>(i + 2)] > 0) {      // 顺子
        ci--;
        c[static_cast<size_t>(i + 1)]--;
        c[static_cast<size_t>(i + 2)]--;
        shape.type[static_cast<size_t>(shape.nSets)] = kSetRun;
        shape.start[static_cast<size_t>(shape.nSets)] = i;
        shape.nSets++;
        enumSets(c, i, need - 1, shape, out);
        shape.nSets--;
        ci++;
        c[static_cast<size_t>(i + 1)]++;
        c[static_cast<size_t>(i + 2)]++;
    }
}

bool isKokushi(const Counts &c) {
    int kinds = 0;
    int total = 0;
    for (int k = 0; k < kKindCount; k++) {
        const int v = c[static_cast<size_t>(k)];
        if (v == 0) {
            continue;
        }
        if (!isYaochu(k)) {
            return false;
        }
        kinds++;
        total += v;
        if (v > 2) {
            return false;
        }
    }
    return kinds == 13 && total == 14;
}

bool allOrphanKinds(const Counts &c) {
    int kinds = 0;
    for (int k = 0; k < kKindCount; k++) {
        if (isYaochu(k) && c[static_cast<size_t>(k)] > 0) {
            kinds++;
        }
    }
    return kinds == 13;
}

/** 为一个「形」加上副露面子，并展开和了牌归属的所有变体（与 Java `addWinVariants` 同序）。 */
void addWinVariants(std::vector<Form> &out, const Shape &shape, int winKind,
                    const std::vector<Meld> &melds) {
    const int pair = shape.pair;
    const int n = shape.nSets;
    const int meldCount = static_cast<int>(melds.size());
    if (n + meldCount != 4) {
        return;
    }
    Form base;
    base.pair = pair;
    base.nSets = n;
    for (int i = 0; i < n; i++) {
        base.setType[static_cast<size_t>(i)] = shape.type[static_cast<size_t>(i)];
        base.setStart[static_cast<size_t>(i)] = shape.start[static_cast<size_t>(i)];
        base.setConcealed[static_cast<size_t>(i)] = true;
        base.setFromMeld[static_cast<size_t>(i)] = false;
    }
    // 和了牌归属：[面子下标(-1 = 雀头), waitType]
    std::array<std::array<int, 2>, 5> assign{};
    int nAssign = 0;
    if (pair == winKind) {
        assign[static_cast<size_t>(nAssign++)] = {-1, kWaitTanki};
    }
    for (int i = 0; i < n; i++) {
        if (base.setType[static_cast<size_t>(i)] == kSetRun) {
            const int s = base.setStart[static_cast<size_t>(i)];
            if (winKind == s) {
                // 缺顺子最小牌：手中剩 s+1,s+2；只有 8,9 才是边张
                assign[static_cast<size_t>(nAssign++)] = {i, (s % 9 == 6) ? kWaitPenchan : kWaitRyanmen};
            } else if (winKind == s + 1) {
                assign[static_cast<size_t>(nAssign++)] = {i, kWaitKanchan};
            } else if (winKind == s + 2) {
                // 缺顺子最大牌：手中剩 s,s+1；只有 1,2 才是边张
                assign[static_cast<size_t>(nAssign++)] = {i, (s % 9 == 0) ? kWaitPenchan : kWaitRyanmen};
            }
        } else if (base.setType[static_cast<size_t>(i)] == kSetTriplet
                   && base.setStart[static_cast<size_t>(i)] == winKind) {
            assign[static_cast<size_t>(nAssign++)] = {i, kWaitShanpon};
        }
    }
    if (nAssign == 0) {
        return;
    }
    for (int a = 0; a < nAssign; a++) {
        Form f = base;
        f.winSet = assign[static_cast<size_t>(a)][0];
        f.waitType = assign[static_cast<size_t>(a)][1];
        int idx = f.nSets;
        for (const Meld &m : melds) {
            f.setType[static_cast<size_t>(idx)] = m.isKan() ? kSetQuad : (m.isRun() ? kSetRun : kSetTriplet);
            f.setStart[static_cast<size_t>(idx)] = m.baseKind();
            f.setConcealed[static_cast<size_t>(idx)] = !m.open();
            f.setFromMeld[static_cast<size_t>(idx)] = true;
            idx++;
        }
        f.nSets = idx;
        out.push_back(f);
    }
}

}  // namespace

std::vector<Form> decompose(const Counts &counts, const std::vector<Meld> &melds, int winKind) {
    std::vector<Form> out;
    const int meldCount = static_cast<int>(melds.size());
    const int need = 4 - meldCount;
    Counts c = counts;
    if (sumOf(c) != need * 3 + 2) {
        return out;
    }
    std::vector<Shape> rawShapes;
    for (int p = 0; p < kKindCount; p++) {
        if (c[static_cast<size_t>(p)] < 2) {
            continue;
        }
        c[static_cast<size_t>(p)] -= 2;
        Shape shape;
        shape.pair = p;
        enumSets(c, 0, need, shape, rawShapes);
        c[static_cast<size_t>(p)] += 2;
    }
    for (const Shape &shape : rawShapes) {
        addWinVariants(out, shape, winKind, melds);
    }
    if (meldCount == 0) {
        // ---- 七对子
        bool ok = true;
        for (int k = 0; k < kKindCount; k++) {
            if (c[static_cast<size_t>(k)] != 0 && c[static_cast<size_t>(k)] != 2) {
                ok = false;
                break;
            }
        }
        if (ok) {
            Form f;
            f.type = kTypeChiitoitsu;
            f.pair = winKind;
            f.nSets = 0;
            f.winSet = -1;
            f.waitType = kWaitTanki;
            out.push_back(f);
        }
        // ---- 国士无双
        if (isKokushi(c)) {
            Form f;
            f.type = kTypeKokushi;
            f.pair = winKind;
            f.nSets = 0;
            Counts pre = c;
            pre[static_cast<size_t>(winKind)]--;
            f.kokushi13 = allOrphanKinds(pre);
            f.winSet = -1;
            f.waitType = kWaitTanki;
            out.push_back(f);
        }
    }
    return out;
}

std::vector<int> agariWaits(const Counts &counts13, int meldCount) {
    std::vector<int> out;
    // ⚠ 非法输入防护：真实牌局里同一牌种最多 4 张，但**语料可以手写**（`trainer settle` 的 `furiten`
    //   行撞到过同种 5 张）。`isAgari` 走查表向听，而表的计数维只覆盖 0..4 → 多一张就**越界崩**
    //   （实测 `0xC0000005`）。Java 的 DFS 对这份输入不崩，但那是"两边都接受非法输入"的另一回事 ——
    //   这里只保证 **C++ 侧不崩**：返回空集，于是对拍会**响亮地失败**而不是把进程挂掉。
    for (int k = 0; k < kKindCount; k++) {
        if (counts13[static_cast<size_t>(k)] > 4) {
            return out;
        }
    }
    Counts c = counts13;
    for (int k = 0; k < kKindCount; k++) {
        if (c[static_cast<size_t>(k)] >= 4) {
            continue;
        }
        c[static_cast<size_t>(k)]++;
        if (isAgari(c, meldCount)) {
            out.push_back(k);
        }
        c[static_cast<size_t>(k)]--;
    }
    return out;
}

}  // namespace trainer
